pkg/ingester/ingester_v2.go

```diff
@@ -455,3 +488,93 @@ func (i *Ingester) closeAllTSDB() {
 	i.userStatesMtx.Unlock()
 	wg.Wait()
 }
+
+/* cleanup walks the TSDB directory for each user to determine if the TSDB should be closed.
+ * If
+ * - The directory is empty of blocks
+ * OR
+ * - The directory has every block marked as shipped
+ * AND
+ * - The TSDB hasn't been written to within the alloted time period
+ * THEN
+ * - Close the TSDB, stop the shipper routine, and delete the directory and it's contents.
+ */
+func (i *Ingester) cleanup() error {
+	ts := time.Now()
+	defer func() {
+		i.metrics.cleanupDuration.Observe(time.Since(ts).Seconds())
+	}()
+
+	userIDs := i.getUsers()
+
+	unshipped, err := unshippedBlocks(i.cfg.TSDBConfig.Dir)
+	if err != nil {
+		return err
+	}
+
+	wg := &sync.WaitGroup{}
+	for _, user := range userIDs {
+
+		// Only try and process users that don't have unshipped blocks
+		if len(unshipped[user]) != 0 {
+			continue
+		}
+
+		// Get the last written time stamp
+		udb := i.getTSDB(user)
+		if udb == nil {
+			level.Warn(util.Logger).Log("msg", "unable to find user tsdb during cleanup", "user", user)
+			continue
+		}
+
+		udb.Lock()
+		last := udb.lastWritten
+		udb.Unlock()
+
+		// If the TSDB is inactive, close and remove it.
+		if ts.Sub(last) > i.cfg.TSDBConfig.MaxStaleAge {
+			wg.Add(1)
+			go func(user string) {
+				defer wg.Done()
+				i.userStatesMtx.Lock()
+				defer i.userStatesMtx.Unlock()
+
+				db, ok := i.TSDBState.dbs[user]
+				if ok {
```

Comment:
What's the case, if `!ok` we want to continue? Looks that if we don't have the TSDB opened, we don't want to continue, so a guard may be better.

Reply:
well if we don't have a TSDB we still want to remove any on disk remnants that may exist. So I think it still makes sense to try and clean those up.

Reply:
This no longer applies dur to recent refactor.